File: nim/cotoha.nim
import class
import json
import httpclient
import strformat
import strutils

class Cotoha:
    var clientId: string
    var clientSecret: string

    proc init*(clientId: string, clientSecret: string) =
        self.clientId = clientId
        self.clientSecret = clientSecret

    proc post(url: string, header: openArray[tuple[key: string, val: string]], param: JsonNode): JsonNode =
        var client = newHttpClient()
        client.headers = newHttpHeaders(header)
        var res: Response = client.request(
            url = url,
            httpMethod = HttpPost,
            body = $param
        )

        if not res.status.contains("200") and not res.status.contains("201"):
            return %*{ "err": fmt"post:status is bad:{res.status}" }

        try:
            return parseJson(res.body)

        except:
            return %*{ "err": fmt"post:parseJson is failed" }

    proc access(): JsonNode =
        return self.post(
            "https://api.ce-cotoha.com/v1/oauth/accesstokens",
            {
                "Content-Type": "application/json",
            },
            %*{
                "grantType": "client_credentials",
                "clientId": self.clientId,
                "clientSecret": self.clientSecret,
            }
        )

    proc common(target: string, param: JsonNode): JsonNode =
        result = parseJson("{}")

        result = self.access()
        if result.hasKey("err"):
            return %*{ "err": fmt"access:{$result[""err""].getStr()}" }

        let access_token: string = $result["access_token"].getStr()

        result = self.post(
            fmt"https://api.ce-cotoha.com/api/dev/nlp/v1/{target}",
            {
            "Content-Type" : "application/json;charset=UTF-8",
            "Authorization" :  fmt"Bearer {access_token}",
            },
            param
        )
        if result.hasKey("err"):
            return %*{ "err": fmt"common:{$result[""err""].getStr()}" }

        if $result["status"] != "0":
            return %*{ "err": fmt"common:api settings are wrong:{$result[""status""]}" }

    proc similarity*(s1: string, s2: string, sen_type: string = "default", dic_type: string = ""): JsonNode =
        return self.common(
            "similarity",
            %*{
                "s1": s1,
                "s2": s2,
                "type": sen_type,
                "dic_type": dic_type,
            }
        )

File: python/cotoha.py
import inspect
import json
import os
import requests
import urllib

class Cotoha():
    def __init__(self, clientId, clientSecret):
        self.__clientId = clientId
        self.__clientSecret = clientSecret

    def __access(self):
        res = requests.post(
            url='https://api.ce-cotoha.com/v1/oauth/accesstokens',
            headers={
                'Content-Type': 'application/json'
            },
            data=json.dumps({
                'grantType': 'client_credentials',
                'clientId': self.__clientId,
                'clientSecret': self.__clientSecret,
            })
        )

        if res.status_code != 200 and res.status_code != 201:
            return { "err" : f"access:status is bad:{str(res.status_code)}" }

        json_data = res.json()

        if json_data["access_token"] == "":
            return { "err" : f"Authentication is failed" }

        return json_data["access_token"]

    def __common(self, param):
        access_token = self.__access()
        if 'err' in access_token:
            return { "err" : f"common:{access_token['err']}" }

        target = inspect.currentframe().f_back.f_code.co_name

        res = requests.post(
            url=f'https://api.ce-cotoha.com/api/dev/nlp/v1/{target}',
            headers={
                'Content-Type': 'application/json;charset=UTF-8',
                'Authorization':  f"Bearer {access_token}",
            },
            data=json.dumps(param)
        )

        if res.status_code != 200 and res.status_code != 201:
            return { "err" : f"common:status is bad:{str(res.status_code)}" }

        json_data = res.json()

        if json_data['status'] != 0:
            return { "err" : f"common:api status is bad:{json_data['status']}" }

        return json_data['result'] 

    def similarity(self, s1, s2, sen_type = "default", dic_type = ""):
        return self.__common(
            {
                "s1": s1,
                "s2": s2,
                "type": sen_type,
                "dic_type": dic_type,
            }
        )
